libnet: create DNS records on sbJoin (if not agent node)

akerouanton opened the pull request:
- Fixes https://github.com/moby/moby/issues/51491
- Related to https://github.com/moby/moby/pull/48971

**- What I did**

Commit https://github.com/moby/moby/commit/a8b9eff902903246aed22a0e49fbc713abcb6a94 removed a call to Network.updateSvcRecord from Network.createEndpoint on the grounds that:

> all callers of Network.createEndpoint follow up with an Endpoint.Join, which also sets up the DNS entry.

However, the original call in Network.createEndpoint was gated by:

```golang
if !n.getController().isSwarmNode() || n.Scope() != scope.Swarm || !n.driverIsMultihost() {
	n.updateSvcRecord(context.WithoutCancel(ctx), ep, true)
}
```

whereas the call in Endpoint.sbJoin() (invoked by Endpoint.Join()) is gated by:

```golang
if !n.getController().isAgent() {
    if !n.getController().isSwarmNode() || n.Scope() != scope.Swarm || !n.driverIsMultihost() {
	    n.updateSvcRecord(context.WithoutCancel(ctx), ep, true)
    }
}
```

As a result, once a node has joined a Swarm cluster, no DNS entries are created for non swarm-scoped networks.

Change the condition used by `sbJoin` to match the original condition used in `createEndpoint`, and add an integration test.

**- Human readable description for the release notes**

```markdown changelog
Fix a bug preventing DNS resolution of containers attached to non swarm-scoped networks once the node has joined a Swarm cluster
```

akerouanton committed:
libnet: create DNS records on sbJoin (if not agent node)

Commit a8b9eff90 removed a call to Network.updateSvcRecord from
Network.createEndpoint on the grounds that:

> all callers of Network.createEndpoint follow up with an Endpoint.Join,
> which also sets up the DNS entry.

However, the original call in Network.createEndpoint was gated by:

```
if !n.getController().isSwarmNode() || n.Scope() != scope.Swarm || !n.driverIsMultihost() {
	n.updateSvcRecord(context.WithoutCancel(ctx), ep, true)
}
```

whereas the call in Endpoint.sbJoin() (invoked by Endpoint.Join()) is
gated by:

```
if !n.getController().isAgent() {
    if !n.getController().isSwarmNode() || n.Scope() != scope.Swarm || !n.driverIsMultihost() {
	    n.updateSvcRecord(context.WithoutCancel(ctx), ep, true)
    }
}
```

As a result, once a node has joined a Swarm cluster, no DNS entries are
created for non swarm-scoped networks.

Change the condition used by `sbJoin` to match the original condition
used in `createEndpoint`.

Signed-off-by: Albin Kerouanton <albin.kerouanton@docker.com>

akerouanton committed:
inte/networking: test DNS resolution for non swarm-scoped nws

Previous commit reverted a faulty change that broke DNS resolution for
non swarm-scoped networks once a node has joined a Swarm cluster.

This commit adds an integration test to verify that we don't break DNS
resolution again.

Signed-off-by: Albin Kerouanton <albin.kerouanton@docker.com>

vvoland reviewed (approved):
SGTM if green

robmry reviewed (approved):
LGTM - thank you.

I think the bug is the `isAgent` check in `sbJoin` (now `populateNetworkResources`). But, as discussed, we don't currently have enough swarm tests or the nerve to change anything.

akerouanton committed:
inte/networking: TestDisableIPv6OnInterface: add '-c1' to ping

If the DNS name still resolves to an IP address, and that address is
assigned to a running container, the ping command will run indefinitely
and the test suite will time out for 10 mins.

This is confusing, as it looks like a daemon hang, or a test suite hang,
whereas it's just a test failure. Add '-c1' to ping to make it return
immediately.

Signed-off-by: Albin Kerouanton <albin.kerouanton@docker.com>
